staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go

```diff
@@ -241,26 +242,36 @@ func (s *WatchServer) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 			}
 			metrics.WatchEvents.WithContext(req.Context()).WithLabelValues(kind.Group, kind.Version, kind.Kind).Inc()
 
-			obj := s.Fixup(event.Object)
-			if err := embeddedEncodeFn(obj, buf); err != nil {
-				// unexpected error
-				utilruntime.HandleError(fmt.Errorf("unable to encode watch object %T: %v", obj, err))
-				return
-			}
+			// Returns an external representation of a watch event that will be then serialized
+			// and sent to the watch client.
+			newOutEvent = func() (runtime.Object, error) {
+				obj := s.Fixup(event.Object)
+				if err := embeddedEncodeFn(obj, buf); err != nil {
+					// unexpected error
+					return nil, fmt.Errorf("unable to encode watch object %T: %v", obj, err)
+				}
 
-			// ContentType is not required here because we are defaulting to the serializer
-			// type
-			unknown.Raw = buf.Bytes()
-			event.Object = &unknown
-			metrics.WatchEventsSizes.WithContext(req.Context()).WithLabelValues(kind.Group, kind.Version, kind.Kind).Observe(float64(len(unknown.Raw)))
+				// ContentType is not required here because we are defaulting to the serializer
+				// type
+				unknown.Raw = buf.Bytes()
+				event.Object = &unknown
 
-			*outEvent = metav1.WatchEvent{}
+				metrics.WatchEventsSizes.WithContext(req.Context()).WithLabelValues(kind.Group, kind.Version, kind.Kind).Observe(float64(len(unknown.Raw)))
 
-			// create the external type directly and encode it.  Clients will only recognize the serialization we provide.
-			// The internal event is being reused, not reallocated so its just a few extra assignments to do it this way
-			// and we get the benefit of using conversion functions which already have to stay in sync
-			*internalEvent = metav1.InternalEvent(event)
-			err := metav1.Convert_v1_InternalEvent_To_v1_WatchEvent(internalEvent, outEvent, nil)
+				if err := metav1.Convert_watch_Event_To_v1_WatchEvent(&event, watchEvent, nil); err != nil {
+					return nil, fmt.Errorf("unable to convert watch object: %v", err)
+				}
+				return watchEvent, nil
+			}
+			var err error
+			if obj, ok := event.Object.(runtime.CacheableObject); ok {
+				outEvent, err = obj.GetCachedEvent(
```

Comment:
I'm afraid outEvent of underlying type metav1.WatchEvent won't be sufficient to properly solve the problem if I read my notes in https://github.com/kubernetes/kubernetes/issues/110146#issue-1242974327 correctly:
* despite sharing the same memory object, we will be still serializing each instance separately in e.Encode(outEvent) which will call json.compact for each watcher separately.

* We need to return here some kind of caching object so that if we pass it to e.Encode it will buffer resulting object.